sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -495,6 +495,23 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Eliminate no-op Projects
     case p @ Project(_, child) if sameOutput(child.output, p.output) => child
 
+    // The column of father project contains not deterministic function
+    // e.g Rand function. father project will be split to two project.
+    case h @ Project(fields, _: LeafNode) if !fields.forall(_.deterministic) =>
```

Comment:
There is no need to split for all LeafNode. if and only if the projectList is non-deterministic for LeafNode's father.